src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java

```diff
@@ -0,0 +1,278 @@
+         * @param maxNumberOfRounds is the maximum number of rounds in simulation (index of last round).
+         * @return this, to provide chaining.
+         */
+        public Builder setMaxNumberOfRounds(int maxNumberOfRounds) {
+            this.maxNumberOfRounds = maxNumberOfRounds;
+            return this;
+        }
+
+        /**
+         * Set the threshold transmission radius.
+         * @param radius is the threshold transmission radius
+         * @return this, to provide chaining.
+         */
+        public Builder setRadius(double radius) {
+            this.radius = radius;
+            return this;
+        }
+
+        /**
+         * Set the awakeness cycle, which is the number of rounds in which every observer must have
+         * an awakeness period.
+         * @param awakenessCycle is the number of rounds to be considered as the cycle.
+         * @return this, to provide chaining.
+         */
+        public Builder setAwakenessCycle(int awakenessCycle) {
+            this.awakenessCycle = awakenessCycle;
+            return this;
+        }
+
+        /**
+         * Set the awakeness duration, which is the number of rounds in which an observer is awake in a
+         * single awakeness cycle.
+         * @param awakenessDuration is the number of rounds to be considered as the duration.
+         * @return this, to provide chaining.
+         */
+        public Builder setAwakenessDuration(int awakenessDuration) {
+            this.awakenessDuration = awakenessDuration;
+            return this;
+        }
+
+        /**
+         * Set the observers' awakeness strategy type, used for generating awakeness strategies for all observers.
+         * @param awakenessStrategyType is the awakeness strategy type for all observers.
+         * @return this, to provide chaining.
+         */
+        public Builder setAwakenessStrategyType(AwakenessStrategyFactory.Type awakenessStrategyType) {
+            this.awakenessStrategyType = awakenessStrategyType;
+            return this;
+        }
+
+        /**
+         * Set the beacons' movement strategy.
+         * @param beaconMovementStrategy is the movement strategy for all beacons.
+         * @return this, to provide chaining.
+         */
+        public Builder setBeaconMovementStrategy(IMovementStrategy beaconMovementStrategy) {
+            this.beaconMovementStrategy = beaconMovementStrategy;
+            return this;
+        }
+
+        /**
+         * Set the observers' movement strategy.
+         * @param observerMovementStrategy is the movement strategy for all observers.
+         * @return this, to provide chaining.
+         */
+        public Builder setObserverMovementStrategy(IMovementStrategy observerMovementStrategy) {
+            this.observerMovementStrategy = observerMovementStrategy;
+            return this;
+        }
+
+        /** Write simulation metadata to db. */
+        abstract void writeSimulationMetadata();
```

Comment:
These method include logic (validation and objects creation) required for to the simulation construction, and are being called inside the `build()` method:

```java
        @Override
        public AbstractSimulation build() {

            // validation of the correctness of the arguments provided using the builder's setters.
            validateArguments(); 

            // board and resolver creation.
            this.realBoard = new RealBoard(this.rowNum, this.colNum);
            this.resolver = new GlobalResolver(this.rowNum, this.colNum);

            // agents creation.
            // (the builder was only given number of beacons / observers, and their strategies).
            initializeBeacons();
            initializeObservers();

            // db interaction, and unique simulation id assignment.
            writeSimulationMetadata();

            // simulation construction.
            return new TracingSimulation(this);
        }
```

The only method not implemented yet is the `writeSimulationMetadata()`, and it should be something like:
```java
        @Override
        public void writeSimulationMetadata() {
            // Write an entity to the db, 
            // and return the unique id the simulation entity was assigned by datastore.
            this.id = DatabaseService.getInstance().writeSimulationMetadata(this); 
        }
```

These methods' logic can be copied into the `build()` method, but I think this will be less readable.

Reply:
Update - the alternative might be moving these methods' logic to the setters:

- `validateArguments()` **is indeed redundant** - this validation can happen inside the setter methods.
- `initializeObservers()` required multiple fields to be already set - `observersNum`, `oberverMovementStrategy` and `awakenessStrategyType`. Thus the observers creation **cannot happen inside any of these setter methods.**
- `initializeBeacons()` is **similar to the previous method**. 
In addition, this shall be an **abstract class** due to the fact that in the stalking simulation the created beacons are instances of the `SwappingBeacon` class.
- `writeSimulationMetadata()` required all fields to be already set. As mentioned above, its implementation is a single-line thus we might **inline it into the `build()` method.**

If you agree with those statements I'll modify these to remove the redundant methods. Please let me know if I am missing something.

@OrpazFishel

Reply:
Final (cross-field) validation indeed must happen during `build()`. I'm fine with that.
The same goes for the complex initialization steps.

As for database accesses - I thought the plan is to move them to separate (adapter) classes? (OK if this is done later.)

Reply:
Will be done later - those adapter classes (`BoardState` and `SimulationMetadata`) can be reviewed at #8.

Thanks!